[Security Solution] Test plan for prebuilt rule customization

dplumlee opened the pull request:
## Summary

Addresses https://github.com/elastic/kibana/issues/202068

Adds test plan for rule customization features related to the milestone 3 prebuilt rule customization epic





<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

banderror commented:
@pborgonovi Please review this one

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+And the rule is unmodified
+When a user edits the rule from the rule edit page to something different than the original version
+Then the rule is successfully updated
+And the ruleSource should be "external"
```
Could we have separate scenarios for `ruleSource` instead of checking it in every test scenario?

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+
+### Assumptions
+
+- Rule package used will have all rule versions present (no missing base versions)
```
```suggestion
- Rule package used will have all previous rule versions present (no missing base versions)
```

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+
+### Terminology
+
+- **Base version**: The version of the rule we ship with the rule package, can be thought of as the "original" version of the rule.
```
```suggestion
- **Base version**: Prebuilt rule asset we ship in the rule package corresponding to the currently installed prebuilt rules.  It represents "original" version of the rule. During prebuilt rules installation prebuilt rule assets data is copied over and becomes an installed prebuilt rule. 
```

nikitaindik replied:
Can we say that it's a prebuilt rule asset whose "version" field matches the "version" of an installed rule? WDYT?

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+
+- **Base version**: The version of the rule we ship with the rule package, can be thought of as the "original" version of the rule.
+
+- **Customized prebuilt rule**: A prebuilt rule that has been changed by the user from the base version of the rule. Also referred to as "Modified" in the UI.
```
```suggestion
- **Customized prebuilt rule**: An installed prebuilt rule that has been changed by the user in the way rule fields semantically differ from the base version. Also referred to as "Modified" in the UI.
```

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+
+- **Customized prebuilt rule**: A prebuilt rule that has been changed by the user from the base version of the rule. Also referred to as "Modified" in the UI.
+
+- **Non-customized prebuilt rule**: A prebuilt rule that has no change from the base version of the rule.
```
```suggestion
- **Non-customized prebuilt rule**: An installed prebuilt rule that has rule fields values matching the base version.
```

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+
+- **Non-customized prebuilt rule**: A prebuilt rule that has no change from the base version of the rule.
+
+- **Custom rules**: A rule created by the user themselves
```
```suggestion
- **Custom rule**: A rule created by the user themselves
```

maximpn reviewed (commented):
@dplumlee Thanks for adding Prebuilt Rules Customization test plan 🙏 

I reviewed the test plan and left suggestions to make the test plan more robust and improve readability.

It looks like this test plan should include the following scenarios as well

- Custom rules shound't have "Modified" badge on rule details page
- Custom rules shouldn't have "Modified" badge on rule management table
- Scenarios covering `ruleSource`
- Users can navigate to rule edit page from prebuilt rule details

Feel free to reach me to discuss it over Zoom in case of questions.

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
````diff
@@ -0,0 +1,281 @@
+
+```Gherkin
+Given a space with at least one non-customized prebuilt rule installed
+When a user edits the <field_name> field to something different than the base version
````
Should we mention that a change that doesn't create a semantic difference, like `1m` to `60s`, should not set `is_customized` to `true`? Or do you think this would be covered by the definition of "change" in the Terminology section?

dplumlee replied:
I'll add further explanation in the terminology section, we already have tests for the semantics you mention in the codebase

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
````diff
@@ -0,0 +1,281 @@
+```Gherkin
+Given a space with at least one customized prebuilt rule installed
+When a user navigates to that rule's detail page
+Then the rule's isCustomized value should be true
````
I think it should be snake case, since it'll come from our API

```suggestion
Then the rule's `is_customized` value should be true
```

nikitaindik commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
````diff
@@ -0,0 +1,281 @@
+And the Modified badge should NOT be present in the table row
+```
+
+#### **Scenario: User should be able to filter by modified rules on the rule updates table**
````
nit: "**in** the rule updates table"

nikitaindik reviewed (commented):
Thanks for the PR, @dplumlee! Overall, the test coverage looks solid. I’ve left a couple of comments for you to consider.

One potential improvement: should we add a scenario to ensure that the “Edit” buttons in the rule details page and rule management table are clickable when the feature flag is enabled?

Additionally, do we want to include a test to verify that non-modifiable fields, such as `version` and `id` are uneditable?”

Looking forward to your thoughts!

dplumlee commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+Then the rule is successfully updated
+And the ruleSource should be "external"
+And isCustomized should be true
+And the "Modified" badge should appear on the rule's detail page
```
We could do that here, we also have tests for the rule management table and update table themselves that have this test in them. Do we want to expand the scope of these tests to include the table as well as the rule details page?

maximpn replied:
Could we have a separate test scenario to assert `Modified` badge appears on `Rules Update` table?

dplumlee commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,281 @@
+And the action is successfully applied to selected rules
+Then rules that have been changed from their base version should have a ruleSource of "external"
+And isCustomized should be true
+And the "Modified" badge should appear on the respective row in the rule management table
```
Again for this one, maybe we can just have one test specifically about the "modified" badge that checks all the locations in one test but for these it seems out of immediate scope wdyt?

dplumlee commented:
@maximpn @nikitaindik I believe I addressed most of your comments or left questions, take a look when you can.

> Scenarios covering ruleSource

@maximpn I think all the additional test cases you mentioned have now been covered but not sure what you meant by this one. I've left a few responses in regard to the `ruleSource` stuff related to your comments but the whole test plan is kind of related to rule source itself.

pborgonovi reviewed (commented):
Hey @dplumlee 
I've left some comments

maximpn commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
````diff
@@ -0,0 +1,354 @@
+| Add rule actions                 |
+```
+
+### Calculating is_customized value
````
```suggestion
### Detecting rule customizations
```

maximpn reviewed (changes requested):
@dplumlee Thanks for addressing my comments 👍 

I believe `ruleSource` and `is_customized` assertions should be removed from test scenarios. Instead of focusing on `ruleSource` and `is_customized` test scenarios should focus on `Modified` badge.

dplumlee commented on x-pack/solutions/security/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/rule_customization.md:
```diff
@@ -0,0 +1,354 @@
+| anomaly_threshold       |
+| new_terms_fields        |
+| history_window_start    |
+| type                    |
```
Moved these to a shared folder

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/12812695257

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/206970">](https://github.com/elastic/kibana/pull/206970)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
